warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ValidComparisonVisitor.java

```diff
@@ -0,0 +1,166 @@
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTModNode node, Object data) {
+        // not concerned with literals for mod nodes
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTEQNode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTNENode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTLTNode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTGTNode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTLENode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTGENode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTERNode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTNRNode node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    @Override
+    public Object visit(ASTAssignment node, Object data) {
+        validateExpression(node, super.visit(node, new MutableInt()));
+        return data;
+    }
+    
+    /* literal visitors */
+    @Override
+    public Object visit(ASTNumberLiteral node, Object data) {
+        return visitLiteral(data);
+    }
+    
+    @Override
+    public Object visit(ASTTrueNode node, Object data) {
+        return visitLiteral(data);
+    }
+    
+    @Override
+    public Object visit(ASTFalseNode node, Object data) {
+        return visitLiteral(data);
+    }
+    
+    @Override
+    public Object visit(ASTNullLiteral node, Object data) {
+        return visitLiteral(data);
+    }
+    
+    @Override
+    public Object visit(ASTStringLiteral node, Object data) {
+        return visitLiteral(data);
+    }
+    
+    private Object visitLiteral(Object data) {
+        if (data instanceof MutableInt) {
+            ((MutableInt) data).increment();
+        }
+        return data;
+    }
+    
+    private void validateExpression(JexlNode node, Object data) {
+        if ((data instanceof MutableInt) && (((MutableInt) data).intValue() > 1)) {
+            throw new DatawaveFatalQueryException("Cannot compare two literals.  Invalid expression: " + JexlStringBuildingVisitor.buildQuery(node));
+        }
+    }
```

Comment:
Doing something like this to pass found identifiers back up to parent nodes fixes the test case I presented...

```suggestion
    private void validateExpression(JexlNode node, Object data, Object parentData) {
        if ((data instanceof MutableInt)) {
            if ((((MutableInt) data).intValue() > 1)) {
                throw new DatawaveFatalQueryException("Cannot compare two literals.  Invalid expression: " + JexlStringBuildingVisitor.buildQuery(node));
            } else if (parentData instanceof MutableInt) {
                ((MutableInt)parentData).add((((MutableInt) data).intValue()));
            }
        }
    }
```

You will obviously have to pass the parent data as the third argument in the various places where this is called.

Reply:
See previous comment.